src/metadata.cpp

```diff
@@ -48,7 +48,21 @@ bool searchRegex(const std::string& regexStr, const std::string& text)
 
 size_t getTextLength(const std::string& utf8EncodedString)
 {
-  return icu::UnicodeString::fromUTF8(utf8EncodedString).length();
+  // For some unknown reason implicite convertion from std::string to icu::StringPiece
+  // is broken on Windows.
+  // Constructors are definde in stringpiece.h as
+  // ```
+  // StringPiece(const std::string& str)
+  //  : ptr_(str.data()), length_(static_cast<int32_t>(str.size())) { }
+  // StringPiece(const char* offset, int32_t len) : ptr_(offset), length_(len) { }
+  // ```
+  // However using the first constructor ends with a corrupted StringPiece (wrong ptr)
+  // and using second one works. Don't ask me why
+  // This is broken
+  // icu::StringPiece stringPiece(utf8EncodedString);
+  // This is not
+  icu::StringPiece stringPiece(utf8EncodedString.data(), static_cast<int32_t>(utf8EncodedString.size()));
+  return icu::UnicodeString::fromUTF8(stringPiece).length();
```

Comment:
The only explanation that I could find is because of the following template constructor of `StringPiece`:

```c++
  /**
   * Constructs from some other implementation of a string piece class, from any
   * C++ record type that has these two methods:
   *
   * \code{.cpp}
   *
   *   struct OtherStringPieceClass {
   *     const char* data();  // or const char8_t*
   *     size_t size();
   *   };
   *
   * \endcode
   *
   * The other string piece class will typically be std::string_view from C++17
   * or absl::string_view from Abseil.
   *
   * Starting with C++20, data() may also return a const char8_t* pointer,
   * as from std::u8string_view.
   *
   * @param str the other string piece
   * @stable ICU 65
   */
  template <typename T,
            typename = typename std::enable_if<
                (std::is_same<decltype(T().data()), const char*>::value
#if defined(__cpp_char8_t)
                    || std::is_same<decltype(T().data()), const char8_t*>::value
#endif
                ) &&
                std::is_same<decltype(T().size()), size_t>::value>::type>
  StringPiece(T str)
      : ptr_(reinterpret_cast<const char*>(str.data())),
        length_(static_cast<int32_t>(str.size())) {}
```

If overload resolution for some reason chooses that constructor over `StringPiece(const std::string& str)` then the string is passed into it by value, and the pointer is bound to the data of a temporary object that is destroyed after the completion of the `StringPiece` constructor. That explanation is valid for an explicit `stringPiece` variable from your commented out example, since the temporary object definitely doesn't survive beyond that point. I am less sure if it can also be valid for the original single liner implementation, as I thought that the temporary objects are required to survive until the end of the full expression in the context of which they were created. But it can be a bug in the compiler (on top of the other bug that chooses the wrong overload).

Reply:
Interesting idea but not conclusive.
I have removed this constructor in the stringpiece header and same error occurs.

> I am less sure if it can also be valid for the original single liner implementation, as I thought that the temporary objects are required to survive until the end of the full expression in the context of which they were created. But it can be a bug in the compiler (on top of the other bug that chooses the wrong overload).

I confirm to you that single liner version is not helping.
A single line "fix" is `return icu::UnicodeString::fromUTF8(utf8EncodedString.data()).length();` but we may truncate binary data containing `\0` so I prefer be explicit about the data size.